pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -203,3 +203,121 @@ func NewPodWatch(client kclient.Interface, namespace, name, resourceVersion stri
 		return obj.(*kapi.Pod)
 	}
 }
+
+func NewContainersReadyCanary(kclient kclient.Interface, timeout time.Duration) *ContainersReadyCanary {
+	return &ContainersReadyCanary{
+		timeout: timeout,
+		podsForDeployment: func(deployment *kapi.ReplicationController) (*kapi.PodList, error) {
+			selector := labels.Set(deployment.Spec.Selector).AsSelector()
+			return kclient.Pods(deployment.Namespace).List(selector, fields.Everything())
+		},
+		getPodStore: func(namespace, name string) (cache.Store, chan struct{}) {
+			sel, _ := fields.ParseSelector("metadata.name=" + name)
+			lw := &deployutil.ListWatcherImpl{
+				ListFunc: func() (runtime.Object, error) {
+					return kclient.Pods(namespace).List(labels.Everything(), sel)
+				},
+				WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+					return kclient.Pods(namespace).Watch(labels.Everything(), sel, resourceVersion)
+				},
+			}
+			store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+			stop := make(chan struct{})
+			cache.NewReflector(lw, &kapi.Pod{}, store, 10*time.Second).RunUntil(stop)
+			return store, stop
+		},
+	}
+}
+
+type ContainersReadyCanary struct {
+	podsForDeployment func(*kapi.ReplicationController) (*kapi.PodList, error)
+	getPodStore       func(namespace, name string) (cache.Store, chan struct{})
+	timeout           time.Duration
+}
+
+func (c *ContainersReadyCanary) Accept(deployment *kapi.ReplicationController) error {
+	// For now, only validate the first replica.
+	if deployment.Spec.Replicas != 1 {
+		return nil
+	}
+
+	// Try and find the pod for the deployment.
+	pods, err := c.podsForDeployment(deployment)
+	if err != nil {
+		return fmt.Errorf("couldn't get pods for deployment %s: %v", deployutil.LabelForDeployment(deployment), err)
+	}
+	if len(pods.Items) == 0 {
+		return fmt.Errorf("no pods found for deployment %s", deployutil.LabelForDeployment(deployment))
+	}
+
+	// If we found multiple, use the first one and log a warning.
+	// TODO: should finding multiple be an error?
+	pod := pods.Items[0]
+	if len(pods.Items) > 0 {
+		glog.Infof("Warning: more than one pod for deployment %s; basing canary check on the first pod '%s'", deployutil.LabelForDeployment(deployment), pod.Name)
+	}
+
+	// Wait for container readiness to be observed once per container in the pod.
+	containersReady := map[string]bool{}
+	for _, container := range pod.Spec.Containers {
+		containersReady[container.Name] = false
+	}
+
+	// Make a pod store to poll and ensure it gets cleaned up.
+	podStore, stopStore := c.getPodStore(pod.Namespace, pod.Name)
```

Comment:
@liggitt 
>> I actually notice that the few places in origin that use watch.Interface don't actually account for any of that.

>Please open issues to track

First I'd like to get @smarterclayton's take. I might be missing why it's okay to not deal with re-establishment.

I need to nail down this design decision ASAP.